src/reconcile/rec_visibility.c

```diff
@@ -414,13 +413,33 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
          * indicate that the value is visible to any timestamp/transaction id ahead of it.
          */
         if (upd->type == WT_UPDATE_TOMBSTONE) {
+tombstone_retry:
             WT_TIME_WINDOW_SET_STOP(select_tw, upd);
             tombstone = upd;
 
             /* Find the update this tombstone applies to. */
             if (!__wt_txn_upd_visible_all(session, upd)) {
                 while (upd->next != NULL && upd->next->txnid == WT_TXN_ABORTED)
                     upd = upd->next;
+                /*
+                 * Handle double tombstone scenario that can happen in the history store when we add
+                 * zero timestamped tombstones to the front of the update chain. In that scenario we
+                 * could have a tombstone with timestamp 0 followed by a tombstone with a non-zero
+                 * timestamp.
+                 *
+                 * We will set the non-zero timestamped tombstone as our currently selected update
+                 * and pin the zero timestamped tombstone in cache. Check that our update isn't the
+                 * same as the update we had further up in order to avoid looping infinitely.
+                 */
+                if (upd->type == WT_UPDATE_TOMBSTONE && upd != tombstone) {
```

Comment:
I don't understand the scenario this is protecting against - could you explain it to me? I'd have thought that a `0` timestamp was globally visible, so does this mean that we end up with a tombstone with a `0` timestamp and a non-zero transaction ID? Why then does that update get pinned into cache?

Reply:
Well the 0 timestamp doesn't make it globally visible as in this change we're adding a txnid that could be not globally visible. 

This scenario aims to protect against the scenario in which we apply that tombstone to a preexisting tombstone. `__wt_rec_upd_select` has specific logic to deal with that zero timestamped, originally globally visible tombstone which isn't anymore as such we need to adjust that logic to deal with the new world.